### ISSUE TITLE
Allow to import public holidays in a popular / reasonable / decent format

### ISSUE THREAD

**Issue by ckunki** [itsallcode/white-rabbit#115]
Potential candidates are 
- flexible definition ala emacs
- ascii or json import format 
- outlook
- ics

Branch: [115_holiday_calculator_plugin](https://github.com/itsallcode/white-rabbit/tree/feature/115_holiday_calculator_plugin)

**Comment by ckunki** [comment]
Some links for inspiration: 
- https://www.google.de/amp/s/www.geeksforgeeks.org/how-to-calculate-the-easter-date-for-a-given-year-using-gauss-algorithm/amp/
- https://de.wikibooks.org/wiki/Algorithmensammlung:_Kalender:_Feiertage
- https://www.gnu.org/software/emacs/manual/html_node/emacs/Holiday-Customizing.html
- https://www.emacswiki.org/emacs/CalendarLocalization#h5o-11

**Comment by kaklakariada** [comment]
The most flexible option could be configuring an iCalendar feed URL providing public holidays, e.g. https://www.mozilla.org/media/caldata/GermanHolidays.ics

Advantages:
* Flexible solution for all countries and religious holidays
* Also works with company specific holidays. A company could publish a custom feed in the intranet.

To simplify configuration, the documentation should point to a collection of public holiday feeds, e.g.
* https://www.thunderbird.net/en-US/calendar/holidays/
* http://icalshare.com/

We could parse the feed e.g. using https://github.com/ical4j/ical4j

**Comment by kaklakariada** [comment]
I think ics is a very flexible and configurable option.

List of public holiday feeds:
* https://www.thunderbird.net/en-US/calendar/holidays/
* http://icalshare.com/

Parser for the iCal format: https://github.com/ical4j/ical4j

What do you think?

**Comment by ckunki** [comment]
Nice. Introduces a dependency. 
Please forgive me - I couldn't resist to attempt a 1000% solution 😎 
see branch: https://github.com/itsallcode/white-rabbit/tree/feature/115_public_holidays

**Comment by kaklakariada** [comment]
Calculating the holidays is of course also nice. But some holidays are not free in all federal states of Germany. And if you want to support other religions it will be complicated ;)

We could implement this via an interface `HolidayService` that returns a list of public holidays for a given month. Plugins can then implement this interface. You can have a calculation-based plugin for German holidays and a plugin for an iCal feed. This gives the user maximum flexibility at the cost of configuration.

**Comment by ckunki** [comment]
Thanks for proposing HolidayService. Sounds good to me and will give maximum flexibility, indeed.
Regarding "only German" - calculation of course is not restricted to German or even Bavarian Holidays.

See : https://www.emacswiki.org/emacs/CalendarLocalization#h5o-11 listing configurations for a variety of countries (incl. Bavaria). 🤣 

I personally would not even restrict the concept "holiday" to a nation, a federal state or a religion. 
Rather I like the idea that holidays are something more personal.
Hence, the selection of a set of holidays could be an individual blend and should of course be configurable to each individual user. This is exactly the concept I plan to implement.
So my idea is to provide generic concepts and functions and enable the user to configure his or her preferred mix of personal holidays with eternal validity.

Of course, calculating Easter might not be sufficient here. 
But having the set of fixed, floating and Easter-based holidays should be a good start already covering a lot of the cases.  
And there are algorithms for most of all the other religious concepts out there. 
Emacs is a great source of inspiration to me.

**Comment by ckunki** [comment]
This brings me to my next question: How should configuration look like?
Where should it be stored?
- ~/.whiterabbit
- ~/.whiterabbit.properties
- user data directory

As configuration is probably a little bit more lengthy (see [115-holidays.txt](https://github.com/itsallcode/white-rabbit/files/6538018/115-holidays.txt)) and will take benefit from some structure elements, properties file does not seem the best option to me. Maybe a dedicated file would be best? This could fit to plugin approach proposed by @kaklakariada. What do you think? 

Should be use XML, json, DSL? XML seems much to verbose and unhandy to me, but so does json. I'd prefer a very short DSL-like format. What do you think?

**Comment by kaklakariada** [comment]
As this will be implemented in a plugin, this won't affect the core. Maybe a file next to `projects.json` in the data directory?

It might be worth to build this as a separate library and include it in WR through a small plugin. Then other Java projects can use the holiday calculation logic.

**Comment by ckunki** [comment]
Sounds good to me, if not to say "fantastic".
Do you have any preference regarding the file format?

**Comment by ckunki** [comment]
I implemented a first parser based on attachment  [115-holidays.txt](https://github.com/itsallcode/white-rabbit/files/6538018/115-holidays.txt)).

**Comment by kaklakariada** [comment]
Regarding the format: I would prefer a DSL like it was used for Emacs. But in Java this would require a compilation step. And adding a script language like groovy is a bit too much.

So a custom text format seems fine. But convert the parsed text into an internal model and then interpret the model in a second step. Separating parsing and interpretation will make it easier to test. And you can change the format later without modifying the Interpreter.

Hope that makes sense ;-)

**Comment by ckunki** [comment]
See new branch [115_holiday_calculator_plugin](https://github.com/itsallcode/white-rabbit/tree/feature/115_holiday_calculator_plugin)

**Comment by ckunki** [comment]
Picking up discussion about scopes and interfaces of WR (incl. WR-HolidayService), WR-Plugin HolidayCalculator and external library holiday-calculator.

### Agreed items

WR-HolidayService (WRHS)
- is required
- should iterate through all plugins implementing PluginFeature `Holidays`  
- should aggregate holidays from all these plugins into WR calendar 

### Open questions 

At the current point in time I am not sure which class shall aggregate holidays for a given month.
For @kaklakariada and me both HolidayProvider as well as HolidayService are OK.
If WRHS is still required and is potentially iterating through multiple plugins anyway then maybe WRHS is the best place for this?
Additionally sorting might be required?

Interface of PluginFeature `Holidays`
- require parameter ModelFactory?  
- return instances implementing interface HolidayProvider.Holiday?

**Comment by ckunki** [comment]
In general I feel some confusion regarding a proper naming of all involved classes. 

I feel the boundaries between WR, plugin and holiday-calculator to be a bit fuzzy or in other words to allow some degree of freedom and flexibility about where to draw the border line.

In some periods I ended up with multiple classes on different layers having identical names and unclear separation of responsibilities or just passing arguments and results back and forth.

On my humble level of development expertise I find identical class names quite confusing and I'd like to avoid them.
I'd love to give each class a perfect name aligned with its responsibilities.

**Comment by ckunki** [comment]
Is it OK, to pass an instance of ModelFactory to HolidayService then? 

This would simplify the code in MonthIndexStorage and enable HolidayService to already return instances of DayData without creating an interface implementation in between in order to just transport the names (and maybe dates) of the holidays.

**Comment by kaklakariada** [comment]
Yes, that's ok. But I would like to avoid passing a ModelFactory to a plugin, because this should be is an internal interface.

**Comment by ckunki** [comment]
Thanks again, @kaklakariada !

In the meantime I took some progress. Code starts to look and feel better now.

I revived interface `HolidayProvider.Holiday` as I felt it to be useful in spite of my initial concerns.

Additionally I renamed interfaces to  `Holidays.HolidayInstance` since I felt that "HolidayProvider" is not a "feature" but more a means to get the *real* feature, which is--simply--the *holidays* themselves.  And as code is crowded by a lot of "Holiday"-This and -that, I wanted to highlight the dedication of the inner interface to *instances* of holidays in opposite to their *definitions*.

Also WR `HolidayService` now looks much better and even simpler taking advantage of the separate class `HolidayAggregator`.
Additionally I plan to rename the `holidaycalculator.*.HolidayService` to `HolidaySet` to avoid classes with identical names.
